pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -18,158 +18,219 @@ package credentials
-	regionURL      string
-	actualProvider *credentialprovider.CachingDockerConfigProvider
+// Provide returns a DockerConfig with credentials from the cache if they are
+// found, or from ECR
+func (p *ecrProvider) Provide(repoToPull string) credentialprovider.DockerConfig {
+	parsed, err := parseRepoURL(repoToPull)
+	if err != nil {
+		klog.Errorf("while parsing ECR repository URL %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+	if cfg, exists := p.getFromCache(parsed); exists {
+		klog.V(3).Infof("Got ECR credentials from cache for %s", parsed.registry)
+		return cfg
+	}
+	cfg, err := p.getFromECR(parsed)
+	if err != nil {
+		klog.Errorf("unable to get credentials from ECR for %s %v", parsed.registry, err)
+	} else {
+		klog.V(3).Infof("Got ECR credentials from ECR API for %s", parsed.registry)
+	}
+	return cfg
 }
 
-var _ credentialprovider.DockerConfigProvider = &lazyEcrProvider{}
+// getFromCache attempts to get credentials from the cache
+func (p *ecrProvider) getFromCache(parsed *parsedURL) (credentialprovider.DockerConfig, bool) {
+	cfg := credentialprovider.DockerConfig{}
 
-// ecrProvider is a DockerConfigProvider that gets and refreshes 12-hour tokens
-// from AWS to access ECR.
-type ecrProvider struct {
-	region    string
-	regionURL string
-	getter    tokenGetter
+	obj, exists, err := p.cache.GetByKey(parsed.registry)
+	if err != nil {
+		klog.V(6).Infof("unable to get credentials from cache for %s %v", parsed.registry, err)
+		return cfg, exists
+	}
+	if exists {
+		entry := obj.(*cacheEntry)
+		cfg[entry.registry] = entry.credentials
+	}
+	return cfg, exists
 }
 
-var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
+// getFromECR gets credentials from ECR since they are not in the cache
+func (p *ecrProvider) getFromECR(parsed *parsedURL) (credentialprovider.DockerConfig, error) {
+	cfg := credentialprovider.DockerConfig{}
+	getter, err := p.getterFactory.GetTokenGetterForRegion(parsed.region)
+	if err != nil {
+		return cfg, err
+	}
+	params := &ecr.GetAuthorizationTokenInput{RegistryIds: []*string{aws.String(parsed.registryID)}}
+	output, err := getter.GetAuthorizationToken(params)
+	if err != nil {
+		return cfg, err
+	}
+	if output == nil {
+		return cfg, errors.New("unable to get ECR token")
+	}
+	if len(output.AuthorizationData) == 0 {
+		return cfg, errors.New("unable to get authorization data")
+	}
+	data := output.AuthorizationData[0]
+	if data.AuthorizationToken == nil {
+		return cfg, errors.New("unable to get authorization token")
+	}
+	if data.ProxyEndpoint == nil {
+		return cfg, errors.New("unable to get proxy endpoint")
+	}
 
-// registryURL has different suffix in AWS China region
-func registryURL(region string) string {
-	dnsSuffix := awsStandardDNSSuffix
-	// deal with aws none standard regions
-	if strings.HasPrefix(region, awsChinaRegionPrefix) {
-		dnsSuffix = awsChinaDNSSuffix
+	decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
```

Comment:
What do you think about having a function that receives the token and returns a cached entry? Would make unit testing the cache entry easier IMO

Reply:
Yes